toolchain/check/context.cpp

```diff
@@ -492,6 +542,96 @@ auto Context::ImplicitAsForArgs(Parse::Node call_parse_node,
   return true;
 }
 
+// Performs a conversion from a tuple to an array type.
+static auto ConvertTupleToArray(Context& context, SemIR::Node tuple_type,
+                                SemIR::Node array_type,
+                                SemIR::TypeId array_type_id,
+                                SemIR::NodeId value_id) -> SemIR::NodeId {
+  auto [array_bound_id, element_type_id] = array_type.GetAsArrayType();
+  auto tuple_elem_types_id = tuple_type.GetAsTupleType();
+  const auto& tuple_elem_types =
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  // Skip back over StubReferences to find if we're being initialized from a
+  // tuple literal.
+  auto value = context.semantics_ir().GetNode(value_id);
+  while (value.kind() == SemIR::NodeKind::StubReference) {
+    value_id = value.GetAsStubReference();
+    value = context.semantics_ir().GetNode(value_id);
+  }
+
+  llvm::ArrayRef<SemIR::NodeId> literal_elems;
+  if (value.kind() == SemIR::NodeKind::TupleValue) {
+    literal_elems =
+        context.semantics_ir().GetNodeBlock(value.GetAsTupleValue());
+  }
+
+  // Check that the tuple is the right size.
+  auto array_bound = context.semantics_ir().GetArrayBoundValue(array_bound_id);
+  if (tuple_elem_types.size() != array_bound) {
+    CARBON_DIAGNOSTIC(
+        ArrayInitFromLiteralArgCountMismatch, Error,
+        "Cannot initialize array of {0} element(s) from {1} initializer(s).",
+        decltype(array_bound), size_t);
```

Comment:
This decltype seems odd to me, is it too difficult to provide a type?

Reply:
Done.